### PR TITLE
[7.x] Casts values from the getOriginal method

### PR DESCRIPTION
**Pull request by aronbeurskens**
# Summary
This merge request fixed the fact that the `getOriginal` doesn't cast the attributes if defined in the $casts array.

# Problem description
The way the `getAttribute` fetches a value differs from `getOriginal`. Take into consideration the following code:

```php
$model = new EloquentModelCastingStub();
$model->intAttribute = '1';
$model->syncOriginal();
$model->intAttribute = '1';
dump($model->getAttribute('intAttribute'), $model->getOriginal('intAttribute'));
```
This will return:

```
1
"1"
```

I've encountered this with SQLite and Oracle DB adapters when fetching a model, changing the value and comparing the original and new values
 
# Expectations
It is expected that both values are returned as integer

# Solution
See pull request, both methods should have the same logic to return the value

### CODE REVIEW

**Review by darielopper — APPROVED** [review]

**Comment by taylorotwell** [comment]
`originalIsEquivalent` is not fully thought through. It is still accepting a `$current` value that is not being used at all. Also, what are the implications on `getDirty`, which is still passing that second value to `originalIsEquivalent`.

**Comment by aronbeurskens** [comment]
In my opinion the `originalIsEquivalent` should not be called with the value to compare to. The method should internally check if a property is modified. For that reason I removed the `$current` property from the method.

File: tests/Integration/Database/EloquentModelStringCastingTest.php [hunk 1]
```diff
@@ -65,15 +65,15 @@ public function testSavingCastedAttributesToDatabase()
             'json_attributes' => ['json_key'=>'json_value'],
             'object_attributes' => ['json_key'=>'json_value'],
         ]);
-        $this->assertSame('{"key1":"value1"}', $model->getOriginal('array_attributes'));
+        $this->assertSame(['key1'=>'value1'], $model->getOriginal('array_attributes'));
         $this->assertSame(['key1'=>'value1'], $model->getAttribute('array_attributes'));
```
Comment (taylorotwell) [inline tests/Integration/Database/EloquentModelStringCastingTest.php]: Why did you change these tests? Aren't they probably protecting against some regression?

Comment (aronbeurskens) [reply tests/Integration/Database/EloquentModelStringCastingTest.php]: I changed those because the test reflect the current situation, but not the desired situation.
If the model is instructed to cast a property to an array, i expect the output of getOriginal to be the same array, not a string.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 2]
```diff
@@ -1154,31 +1134,28 @@ public function getChanges()
...
     {
         if (! array_key_exists($key, $this->original)) {
             return false;
         }
 
+        $attribute = $this->getAttribute($key);
         $original = $this->getOriginal($key);
 
-        if ($current === $original) {
+        if ($attribute === $original) {
```
Comment (rbuzzibarrios) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: I think is better check fail case first. 
```
if (is_null($attribute)) {
    return false;
} elseif ($attribute === $original) { 
    return true;
} elseif ($this->isDateAttribute($key)) {
    return $this->fromDateTime($attribute) === $this->fromDateTime($original);
}
```
originalIsEquivalent must return `true` if `$original` is null and `$attribute` is null too ?

File: tests/Integration/Database/EloquentModelStringCastingTest.php [hunk 3]
```diff
@@ -65,15 +65,15 @@ public function testSavingCastedAttributesToDatabase()
...
 
-        $this->assertSame('{"json_key":"json_value"}', $model->getOriginal('json_attributes'));
+        $this->assertSame(['json_key'=>'json_value'], $model->getOriginal('json_attributes'));
```
Comment (rbuzzibarrios) [inline tests/Integration/Database/EloquentModelStringCastingTest.php]:
```suggestion
        $this->assertSame(['json_key' => 'json_value'], $model->getOriginal('json_attributes'));
```

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php (hunk 2 above)
Comment (aronbeurskens) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: I agree, but it is already existing code and beyond the scope of this PR. I'm happy to refactor this method as well, but I think it should be in a different PR.

Comment (rbuzzibarrios) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: I agree.

**Review by rbuzzibarrios — APPROVED** [review]

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 4]
```diff
@@ -1250,4 +1227,38 @@ protected static function getMutatorMethods($class)
...
     }
+
+    /**
+     * Get the plain value (no relation) of a model respecting the models mutator, casts and dates property.
+     *
+     * @param  string  $key
+     * @param  mixed $value
+     * @return mixed|array
+     */
+    private function getValue($key, $value)
```
Comment (mfn) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: This shouldn't be `private`; the framework never(*) uses this visibility as it prevents customizations.

This should have been `protected`.

(*) yeah, they're like 10 references in the whole codebase

Comment (aronbeurskens) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: @mfn is there a guideline / doc where this (no use of privates) is defined? Maybe with other guidelines to follow?

Comment (mfn) [reply src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: I'm not aware.

The Frameworks thousands like of code are the guidelines I guess.

I've had the pleasure to customize a lot of Laravel internals and that's why I know there are almost no private methods compared to that everything else is protected; which is the reason it's so easy to extend.
